File: 104.maximum-depth-of-binary-tree.py
#
# @lc app=leetcode id=104 lang=python3
#
# [104] Maximum Depth of Binary Tree
#
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
'''
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        count=1
        return count+=max(self.maxDepth(root.left),self.maxDepth(root.right))
'''
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        return max(self.maxDepth(root.left),self.maxDepth(root.right))+1
        '''做法和c++的方法类似

File: 104.maximum-depth-of-binary-tree.cpp
/*
 * @lc app=leetcode id=104 lang=cpp
 *
 * [104] Maximum Depth of Binary Tree
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (root == NULL) return 0;
        int count = 1;
        return count += std::max(maxDepth(root->left), maxDepth(root->right));
    }
};

/* 递归 递归的思想是 加入一个比较过程，把每条子树的长度都算出来，然后
两两比较 最后找到最长的一个子树